lib/fsi.go

```diff
@@ -17,314 +17,402 @@ import (
 	"github.com/qri-io/qri/repo"
 )
 
-// FSIMethods encapsulates filesystem integrations methods
+// FSIMethods groups together methods for FSI
 type FSIMethods struct {
 	inst *Instance
 }
 
-// NewFSIMethods creates a fsi handle from an instance
-func NewFSIMethods(inst *Instance) *FSIMethods {
-	return &FSIMethods{inst: inst}
+// Name returns the name of this method group
+func (m *FSIMethods) Name() string {
+	return "fsi"
 }
 
-// CoreRequestsName specifies this is a fsi handle
-func (m FSIMethods) CoreRequestsName() string { return "fsi" }
+// Filesys returns the FSIMethods that Instance has registered
+func (inst *Instance) Filesys() *FSIMethods {
+	return &FSIMethods{inst: inst}
+}
 
-// LinkParams encapsulate parameters to the link method
+// LinkParams encapsulate parameters for linked datasets
 type LinkParams struct {
-	Dir string
-	Ref string
+	Dir    string
+	Refstr string
+}
+
+// FSIWriteParams encapsultes arguments for writing to an FSI-linked directory
+type FSIWriteParams struct {
+	Refstr string
+	Ds     *dataset.Dataset
 }
 
+// RestoreParams provides parameters to the restore method.
+type RestoreParams struct {
+	Dir       string
+	Refstr    string
+	Path      string
+	Component string
+}
+
+// InitDatasetParams proxies parameters to initialization
+type InitDatasetParams = fsi.InitParams
+
+// StatusItem is an alias for an fsi.StatusItem
+type StatusItem = fsi.StatusItem
+
 // CreateLink creates a connection between a working drirectory and a dataset history
-func (m *FSIMethods) CreateLink(p *LinkParams, res *dsref.VersionInfo) (err error) {
-	// absolutize path name
-	path, err := filepath.Abs(p.Dir)
+func (m *FSIMethods) CreateLink(ctx context.Context, p *LinkParams) (*dsref.VersionInfo, error) {
+	got, err := m.inst.Dispatch(ctx, m.Name()+".createlink", p)
+	if res, ok := got.(*dsref.VersionInfo); ok {
+		return res, err
+	}
+	return nil, dispatchReturnError(got, err)
```

Comment:
the `m.Name()+".createLink"` feels a bit too fragile. I'd prefer a private instance method defined in `dispatch.go`:

```go
func (inst *Instance) dispatchMethodName(m Methods, name string) string 
```

This mainly helps us identify all of these method name constructions for later refactoring

Reply:
I agree it's not ideal. Not sure what `dispatchMethodName` is doing, is the input the current short method name (like "createLink") while the output is the fully qualified name?

Reply:
ah my bad I totally should have filled in the method body: 

```go
func dispatchMethodName(m Methods, name string) string  {
  return fmt.Sprintf("%s.%s", m.Name(), name)
}
```

filling it out here it totally doesn't need the `inst` receiver. The big thing this is doing is only accepting (and utilizing) the `Name` method on the `Methods` interface to force the convention to agree